### PR TITLE
add tolerance to tools._golden_sect_Dataframe

### LINKED ISSUE #1087: Apparent numerical instability in I_mp calculation using PVsyst model
**Describe the bug**

I used these parameters in `pvlib.pvsystem.calcparams_pvsyst()` in order to calculate `I_mp` vs. `T` using `pvlib.pvsystem.singlediode()` with `effective_irradiance` fixed at 1000 W/m^2 and `temp_cell` having 1001 values ranging from 15 to 50 degC:

`{'alpha_sc': 0.006, 'gamma_ref': 1.009, 'mu_gamma': -0.0005, 'I_L_ref': 13.429, 'I_o_ref': 3.719506010004821e-11, 'R_sh_ref': 800.0, 'R_sh_0': 3200.0, 'R_s': 0.187, 'cells_in_series': 72, 'R_sh_exp': 5.5, 'EgRef': 1.121, 'irrad_ref': 1000, 'temp_ref': 25}`

My purpose was to investigate the temperature coefficient of `I_mp`, and I got the following result, which appears to suffer from a numeric instability:

image

For comparison, the corresponding `V_mp` vs. `T` plot:

image

**To Reproduce**

Run the above calculations using the parameters provided.

**Expected behavior**

Better numerical stability in `I_mp` vs. `T`.

**Screenshots**

See above.

**Versions:**

 - ``pvlib.__version__``: 0.8.0
 - ``numpy.__version__``: 1.19.2
 - ``scipy.__version__``: 1.5.2
 - ``pandas.__version__``: 1.1.3
 - python: 3.8.5

**Additional context**

I was going to attempt a numerical computation of the temperature coefficient of `I_mp` for a model translation to the SAPM. I have seen reports from CFV in which this coefficient is actually negative, and I have computed it alternately using the `P_mp` and `V_mp` temperature coefficients, and gotten a negative value for this particular PV module. Despite the apparent numerical instability in the above plot, it still suggests that the coefficient should be positive, not negative. Perhaps I am missing something here?

Also, I have not dug deep enough to figure out if the underlying issue is in `pvlib.pvsystem.singlediode()`.

### PR DESCRIPTION
**Pull request by cwhanse**
- [x] Closes #1087
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - ~~[ ] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.~~
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Sticking with `tools._golden_sect_DataFrame` because it vectorizes single-variable optimization, which looks clumsy (if possible at all) with existing `scipy` tools.  Absolute tolerance was hardwired at 0.01V. Setting a default absolute tolerance of 1e-8V takes care of the reported instability.

### CODE REVIEW

**Comment by wholmgren** [comment]
> Sticking with tools._golden_sect_DataFrame because it vectorizes single-variable optimization, which looks clumsy (if possible at all) with existing scipy tools.

If I recall correctly, @mikofski contributed a vectorized newton optimizer to scipy in large part so that we could remove this function.

**Comment by cwhanse** [comment]
He did, and it's in scipy now, thanks for that reminder. I'll have to think a bit about ways to get an initial estimate of Vmp that minimizes risk of divergence. That's why the bisection algorithm was originally in PVLib for Matlab, which led to the golden section function.

It also appears that at least some of the previous "known" solutions for single diode equations using the 'lambertw' method were not converged.

**Comment by mikofski** [comment]
I added 2 SciPy methods, the vectorized newton method which uses NumPy, and a cythonized api to the bounded search methods, with examples to call them with cythonize `prange` and `nogil` so they're two different things. Altho easier, the newton method would be useful here, instead I'd recommend the cythonized BrentQ

**Comment by mikofski** [comment]
[`cython_optimize_brentq_benchmark.ipynb`](https://gist.github.com/mikofski/ac30065073d0d32d6ea3569f6e24e5ec)

**Comment by mikofski** [comment]
[Toms 748](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.toms748.html) is another bounded search algorithm but idk if it's vectorized

**Comment by cwhanse** [comment]
This is ready for review.

I recalculated the IV curves used for several tests to add precision. I believe the old expected results were done by me, in Matlab, before we [improved convergence](https://github.com/sandialabs/MATLAB_PV_LIB/pull/23) in the PV_Lib Matlab code.

**Comment by wholmgren** [comment]
> before we improved convergence in the PV_Lib Matlab code.

I'm a little confused... the linked PR adds a warning to the user that iteration limit was exceeded, but you're removing it in this PR. Maybe you meant to point to a different PR? Still wondering if removing that exception is a good idea though. Perhaps a warning would be better than an exception and maybe more consistent with other solvers.

**Comment by cwhanse** [comment]
> I'm a little confused

Understandable. That work in PV_LIB for Matlab wasn't done via PRs, now that I look again at the one I linked.

Theoretically, the exit criterion (`err < atol`) will be satisfied before `interations` reaches the computed `iterlimit`. I suppose some very pathological `func` could be engineered to make this fail, but I couldn't find one so I remove the exception to satisfy codecov. I'm OK putting it back, either as an exception or warning, but that line won't be hit by a test.

**Comment by wholmgren** [comment]
> but that line won't be hit by a test.

You can add [`# pragma: no cover`](https://coverage.readthedocs.io/en/coverage-4.3.3/excluding.html) to prevent the coverage tools from complaining.

I'm ok moving forward without the check if others are too.

File: pvlib/tools.py [hunk 1]
```diff
@@ -277,51 +277,57 @@ def _build_args(keys, input_dict, dict_name):
...
         of inputs to the function to be optimized.
         Each row should represent an independent optimization.
 
     VL: float
-        Lower bound of the optimization
+        Lower bound on voltage for the optimization
 
-    VH: float
-        Upper bound of the optimization
+    VH: array-like
```
Comment (kanderso-nrel) [inline pvlib/tools.py:298]: Should `VL` be array-like as well?

File: pvlib/tools.py [hunk 2]
```diff
@@ -277,51 +277,57 @@ def _build_args(keys, input_dict, dict_name):
...
 
     Parameters
     ----------
-    params : dict
-        Dictionary containing scalars or arrays
+    params : dict or Dataframe
+        Parameters for the IV curve(s) where MPP will be found. Must contain
+        keys: 'r_sh', 'r_s', 'nNsVth', 'i_0', 'i_l'
         of inputs to the function to be optimized.
         Each row should represent an independent optimization.
```
Comment (kanderso-nrel) [inline pvlib/tools.py:293]: Should these two lines be removed?  It doesn't flow perfectly with the new text

File: docs/sphinx/source/whatsnew/v0.9.1.rst [hunk 3]
```diff
@@ -24,6 +24,8 @@ Bug fixes
   and ``surface_azimuth`` inputs caused an error (:issue:`1127`, :issue:`1332`, :pull:`1361`) 
 * Changed the metadata entry for the wind speed unit to "Wind Speed Units" in
   the PSM3 iotools function (:pull:`1375`)
+* Improved convergence to maximum power point for :py:func:`pvlib.pvsystem.singlediode`
+  (:issue:`1087`, :pull:`1089`)
```
Comment (kanderso-nrel) [inline docs/sphinx/source/whatsnew/v0.9.1.rst:28]: It would be nice to describe how big a difference to expect from this change.

**Review by kanderso-nrel — COMMENTED** [review]
> I'm ok moving forward without the check if others are too.

I'm not familiar enough with what's happening here to vote.  I don't object, anyway.

File: pvlib/tools.py (hunk 2 above)
Comment (cwhanse) [reply pvlib/tools.py:293]: I rewrote that parameter description, most of that text is leftover from a previous version.

File: docs/sphinx/source/whatsnew/v0.9.1.rst [hunk 4]
```diff
@@ -24,6 +24,10 @@ Bug fixes
   and ``surface_azimuth`` inputs caused an error (:issue:`1127`, :issue:`1332`, :pull:`1361`) 
 * Changed the metadata entry for the wind speed unit to "Wind Speed Units" in
   the PSM3 iotools function (:pull:`1375`)
+* Improved convergence when determining the maximum power point using
+  for :py:func:`pvlib.pvsystem.singlediode` with `method='lambertw'`. Tolerance
```
Comment (wholmgren) [inline docs/sphinx/source/whatsnew/v0.9.1.rst:28]:
```suggestion
  for :py:func:`pvlib.pvsystem.singlediode` with ``method='lambertw'``. Tolerance
```

**Review by wholmgren — APPROVED** [review]
Shall we merge?

I am wondering if there's a meaningful difference in execution time. If so, might be worth noting in the what's new.

**Comment by cwhanse** [comment]
For the golden section function, the old default tolerance was 1e-2. With the new default of 1e-8, convergence takes 3x longer when applied to the test function. That's probably typical of what will be seen when solving single diode equations.

```
from pvlib.tools import _golden_sect_DataFrame
import timeit

def _obj_test_golden_sect(params, loc):
    return params[loc] * (1. - params['c'] * params[loc]**params['n'])


def test__golden_sect_DataFrame_atol(atol):
    params = {'c': 0.2, 'n': 0.3}
    expected = 89.14332727531685
    v, x = _golden_sect_DataFrame(
        params, 0., 100., _obj_test_golden_sect, atol=atol)


t8 = timeit.timeit('test__golden_sect_DataFrame_atol(1e-8)', number=10000, globals=globals())
t2 = timeit.timeit('test__golden_sect_DataFrame_atol(1e-2)', number=10000, globals=globals())
```

`t8` = 6.4s, and `t2`=2.5s.

File: pvlib/tools.py [hunk 5]
```diff
@@ -277,51 +277,61 @@ def _build_args(keys, input_dict, dict_name):
...
-    df['VH'] = VH
-    df['VL'] = VL
+    df['VH'] = upper
+    df['VL'] = lower
 
-    errflag = True
+    converged = False
     iterations = 0
+    iterlimit = 1 + np.max(
+        np.trunc(np.log(atol / (df['VH'] - df['VL'])) / np.log(phim1)))
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/tools.py:330]: I agree. But handling `nan` in the bounds and where returned by `func` is a bit more involved. Currently, I think the optimization would run to its iteration limit.

**Pull request merged**